src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs

```diff
@@ -5986,29 +5970,42 @@ private static void AnalyzeSnapinDomainHelper()
         }
 #endif
 
-        // Analyzes a script module implementation for its exports.
-        private static Dictionary<string, PSModuleInfo> s_scriptAnalysisCache = new Dictionary<string, PSModuleInfo>();
+        /// <summary>
+        /// Cache of module infos from analyzed script modules.
+        /// </summary>
+        private static ConcurrentDictionary<string, ScriptModuleCacheEntry> s_scriptAnalysisCache = new ConcurrentDictionary<string, ScriptModuleCacheEntry>();
 
-        private PSModuleInfo AnalyzeScriptFile(string filename, bool force, ExecutionContext context)
+        /// <summary>
+        /// Analyzes a script module implementation for its exports.
+        /// </summary>
+        /// <param name="filePath">the path to the module to analyze</param>
+        /// <param name="force">if true, ignore the AnalysisCache and re-analyze the script module</param>
+        /// <param name="context">the execution context the script module belongs to</param>
+        /// <returns></returns>
+        private PSModuleInfo AnalyzeScriptFile(string filePath, bool force, ExecutionContext context)
         {
-            // We need to return a cloned version here.
-            // This is because the Get-Module -List -All modifies the returned module info and we do not want the original one changed.
-            PSModuleInfo module = null;
+            // Look up the module to see if we have it cached
+            ScriptModuleCacheEntry cacheEntry = null;
+            s_scriptAnalysisCache.TryGetValue(filePath, out cacheEntry);
 
-            lock (s_lockObject)
+            // Now check the file to see if its changed since we cached it
+            DateTime lastModuleWriteTimeUtc;
+            TryGetFileLastWriteTimeUtc(filePath, out lastModuleWriteTimeUtc);
+
+            // Return here if the cache is up to date
+            if (cacheEntry?.ModuleInfo != null && lastModuleWriteTimeUtc == cacheEntry.LastFileWriteTimeUtc)
```

Comment:
About `TryGetFileLastWriteTimeUtc` see my comment below. We could do:
```c#
var lastModuleWriteTimeUtc = GetFileLastWriteTimeUtc(filePath)
if (s_scriptAnalysisCache.TryGetValue(filePath, out ScriptModuleCacheEntry cacheEntry) && cacheEntry.ModuleInfo != null && lastModuleWriteTimeUtc == cacheEntry.LastFileWriteTimeUtc)
{

}
```

And I wonder why here `ModuleInfo` could be null?